### PR TITLE
[2.6] Rebase multi-tenancy

### PR DESCRIPTION
**Pull request by yannuil**
This pull request rebases all multi-tenancy and no-cluster-rbac changes. 

I have also attempted to integrate our `xns-informer` into the upstream's informer factory. which was introduced in https://github.com/istio/istio/pull/44899.

### CODE REVIEW

File: pilot/pkg/bootstrap/configcontroller.go [hunk 1]
```diff
@@ -140,15 +140,8 @@ func (s *Server) initK8SConfigStore(args *PilotArgs) error {
...
-	var WaitForCRD func(class schema.GroupVersionResource, stop <-chan struct{}) bool
-	if s.kubeClient.CrdWatcher() == nil {
-		WaitForCRD = func(class schema.GroupVersionResource, stop <-chan struct{}) bool {
-			return true
-		}
-	} else {
-		WaitForCRD = func(class schema.GroupVersionResource, stop <-chan struct{}) bool {
-			return s.kubeClient.CrdWatcher().WaitForCRD(class, stop)
-		}
+	WaitForCRD := func(class schema.GroupVersionResource, stop <-chan struct{}) bool {
```
Comment (rcernich) [inline pilot/pkg/bootstrap/configcontroller.go]: I thought the reason we had this code was because istiod does not have permission to watch CRD resources.  Is this still true (or am I mistaken)?

Comment (jewertow) [reply pilot/pkg/bootstrap/configcontroller.go]: It looks like a refactor, so I suggest to revert this change to not introduce unnecessary patches.

File: pilot/pkg/config/kube/crdclient/client.go
Comment (jewertow) [inline pilot/pkg/config/kube/crdclient/client.go:1]: Didn't you forget to add `EnableCRDScan bool` to `Option`? Or we no longer need it?

**Review by jewertow — COMMENTED** [review]
`test.integration.servicemesh.gateway.kube` fails, because the deployment controller can't evaluate template:
```
2024-03-11T21:10:25.130122Z	info	gateway	reconciling	gateway=app-2-96009/gateway
2024-03-11T21:10:25.132193Z	error	controllers	error handling app-2-96009/gateway, retrying (retry count: 1): failed to render template: template: inject:303:11: executing "inject" at <.ServiceType>: can't evaluate field ServiceType in type gateway.derivedInput	controller=gateway deployment
2024-03-11T21:10:25.138139Z	info	gateway	reconciling	gateway=app-2-96009/gateway
...
2024-03-11T21:10:25.217926Z	error	controllers	error handling app-2-96009/gateway, and retry budget exceeded: failed to render template: template: inject:303:11: executing "inject" at <.ServiceType>: can't evaluate field ServiceType in type gateway.derivedInput	controller=gateway deployment
```

File: pkg/kube/namespace/filter.go [hunk 2]
```diff
@@ -266,3 +267,81 @@ func (d *discoveryNamespacesFilter) isSelected(labels labels.Set) bool {
...
+// FilterNamespace is not implemented, because Maistra does not filter namespaces;
+// this function should never be called if memberRollName is specified, so it's safe to call panic.
+func (d *maistraDiscoveryNamespacesFilter) FilterNamespace(nsMeta metav1.ObjectMeta) bool {
+	panic("Not implemented")
+}
+
+// SelectorsChanged is not implemented, because Maistra does not filter namespaces by selectors;
+// this function should never be called if memberRollName is specified, so it's safe to call panic.
+func (d *maistraDiscoveryNamespacesFilter) SelectorsChanged(discoverySelectors []*metav1.LabelSelector) {
+	log.Warn("Discovery namespaces filter not supported in multi-tenancy. Ignore selector changes.")
```
Comment (jewertow) [inline pkg/kube/namespace/filter.go]: In the previous releases we called `panic` here, because this function could not be reached when discovery selectors were configured. Why does that change? Discovery selectors work only in the cluster-wide mode and then xns-informer should not be used.

Comment (yannuil) [reply pkg/kube/namespace/filter.go]: This [line](https://github.com/istio/istio/blob/release-1.20/pilot/pkg/serviceregistry/kube/controller/discoverycontrollers.go#L76) will be triggered during the initialization.

File: pilot/pkg/bootstrap/configcontroller.go (hunk 1 above)
Comment (yannuil) [reply pilot/pkg/bootstrap/configcontroller.go]: Rob, Yes. In the upstream, they stop scanning the CRD at starting the server, and they also introduced a `DelayedInformer`, which works closely with the `CrdWatcher`. I move the CRD related logic into `CrdWatcher` in this [commit](https://github.com/maistra/istio/pull/920/commits/e94bf1c1995c3d1069943c7bc775ebc09536b9b2).

File: pilot/pkg/config/kube/crdclient/client.go
Comment (yannuil) [reply pilot/pkg/config/kube/crdclient/client.go:1]: In the upstream, they stop the CRD scanning.

File: cni/pkg/ebpf/server/ambient_redirect_bpf.o
Comment (dgn) [inline cni/pkg/ebpf/server/ambient_redirect_bpf.o:1]: accident?

File: tests/integration/servicemesh/maistra/maistra.go [hunk 3]
```diff
@@ -0,0 +1,300 @@
...
+    k8s:
+      overlays:
+      - apiVersion: apps/v1
+        kind: Deployment
+        name: istiod-%[2]s
+        patches:
+        - path: spec.template.spec.containers.[name:discovery].args[-1]
+          value: "--memberRollName=default"
+        - path: spec.template.spec.containers.[name:discovery].args[-1]
+          value: "--enableCRDScan=false"
```
Comment (dgn) [inline tests/integration/servicemesh/maistra/maistra.go]: this flag can be removed now, right?

File: pkg/kube/namespace/filter.go (hunk 2 above)
Comment (jewertow) [reply pkg/kube/namespace/filter.go]: Ok, thanks

File: pilot/pkg/config/kube/crdclient/client.go
Comment (jewertow) [reply pilot/pkg/config/kube/crdclient/client.go:1]: Ok, thanks

File: cni/pkg/ebpf/server/ambient_redirect_bpf.o
Comment (yannuil) [reply cni/pkg/ebpf/server/ambient_redirect_bpf.o:1]: No, it was not by accident. The version of libbpf in our builder image is different from what the upstream uses. The file was generated after running `make gen`.

File: tests/integration/servicemesh/maistra/maistra.go (hunk 3 above)
Comment (yannuil) [reply tests/integration/servicemesh/maistra/maistra.go]: Done

File: go.mod [hunk 4]
```diff
@@ -247,5 +250,9 @@ require (
 	sigs.k8s.io/json v0.0.0-20221116044647-bc3834ca7abd // indirect
 	sigs.k8s.io/kustomize/api v0.13.5-0.20230601165947-6ce0bf390ce3 // indirect
 	sigs.k8s.io/kustomize/kyaml v0.14.3 // indirect
-	sigs.k8s.io/structured-merge-diff/v4 v4.3.0 // indirect
+	sigs.k8s.io/structured-merge-diff/v4 v4.4.1 // indirect
 )
+
+replace github.com/maistra/xns-informer => github.com/yannuil/maistra-xns-informer v0.0.0-20240129091448-e34587be7ad1
+
+replace maistra.io/api => github.com/yannuil/maistra-api v0.0.0-20240129092050-f6b2453efd2d
```
Comment (jewertow) [inline go.mod]: What about PRs to xns-informer and maistra-api?

File: pkg/kube/namespace/filter.go (hunk 2 above)
Comment (jewertow) [reply pkg/kube/namespace/filter.go]: Instead of logging this warning, I suggest to change [initMeshWatcherHandler](https://github.com/istio/istio/blob/release-1.20/pilot/pkg/serviceregistry/kube/controller/discoverycontrollers.go#L74) as follows and call `panic` here, as it was in previous releases:
```
func (c *Controller) initMeshWatcherHandler(meshWatcher mesh.Watcher, discoveryNamespacesFilter filter.DiscoveryNamespacesFilter) {
	meshWatcher.AddMeshHandler(func() {
		if c.client.GetMemberRollController() != nil {
			discoveryNamespacesFilter.SelectorsChanged(meshWatcher.Mesh().GetDiscoverySelectors())
		}
	})
}
```

File: go.mod
Comment (jewertow) [inline go.mod:1]: I guess you upgraded `github.com/go-logr/logr`, `golang.org/x/sys` and indirect dependencies accidentally. If that's true, could you revert these changes?

File: go.mod [hunk 5]
```diff
@@ -108,8 +109,9 @@ require (
 	gopkg.in/yaml.v2 v2.4.0
 	gopkg.in/yaml.v3 v3.0.1
 	helm.sh/helm/v3 v3.12.3
-	istio.io/api v1.20.0-rc.0
-	istio.io/client-go v1.20.0-rc.0
+	istio.io/api v1.20.2-0.20231213020515-8655fab91d5d
+	istio.io/client-go v1.20.2-0.20231213021015-d9f0411d8b67
+	istio.io/pkg v0.0.0-20231221211216-7635388a563e
```
Comment (jewertow) [inline go.mod]: Why is this package needed?

File: pilot/pkg/bootstrap/configcontroller.go (hunk 1 above)
Comment (jewertow) [reply pilot/pkg/bootstrap/configcontroller.go]: Sorry but I will push to not add any refactor to the patched code. I put a lot of efforts into minimizing patches while working on 2.3 and 2.5.

File: pilot/pkg/config/kube/crdclient/client.go [hunk 6]
```diff
@@ -359,6 +365,7 @@ func (cl *Client) addCRD(name string) {
 	if s.IsBuiltin() {
 		kc = kclient.NewUntypedInformer(cl.client, gvr, filter)
 	} else {
+		cl.logger.Infof("delayedInformer for %v", gvr)
```
Comment (jewertow) [inline pilot/pkg/config/kube/crdclient/client.go]: Is this needed or is it a leftover of debugging?

**Review by jewertow — COMMENTED** [review]
Could you properly squash related commits to keep them well described and organized like [here](https://github.com/maistra/istio/pull/814/commits)?

File: pilot/pkg/config/kube/crdclient/client.go (hunk 6 above)
Comment (yannuil) [reply pilot/pkg/config/kube/crdclient/client.go]: Leftover. Thanks for spotting that and I will remove it.

File: pilot/pkg/bootstrap/configcontroller.go (hunk 1 above)
Comment (yannuil) [reply pilot/pkg/bootstrap/configcontroller.go]: This is a mistake. I was trying to revert my changes as it was not needed anymore.

File: go.mod (hunk 4 above)
Comment (yannuil) [reply go.mod]: I will create these PRs first.

File: go.mod (hunk 5 above)
Comment (yannuil) [reply go.mod]: I don't think I update this intentionally. I will try to revert this changes.

**Comment by yannuil** [comment]
I am organizing the commits into ones like the 2.5 PR

File: pkg/kube/namespace/filter.go (hunk 2 above)
Comment (yannuil) [reply pkg/kube/namespace/filter.go]: Done

Comment (yannuil) [reply pkg/kube/namespace/filter.go]: I had to revert the suggested changes as it fails the unit test locally. I will investigate.

Comment (yannuil) [reply pkg/kube/namespace/filter.go]: Found the bug. We should call `SelectedChange` when `c.client.GetMemberRollController() == nil`, not `!=`.

**Review by jewertow — APPROVED** [review]
